spec/acceptance/suites/default/redis_multi_instances_one_host_spec.rb

```diff
@@ -23,67 +23,55 @@
                  'redis'
                end
 
-  it 'runs successfully' do
-    pp = <<-EOS
-    $listening_ports = #{instances}
+  include_examples 'an idempotent resource' do
+    let(:manifest) do
+      <<-PUPPET
+      $listening_ports = #{instances}
 
-    class { '::redis':
-      default_install => false,
-      service_enable  => false,
-      service_ensure  => 'stopped',
-    }
-
-    $listening_ports.each |$port| {
-      $port_string = sprintf('%d',$port)
-      redis::instance { $port_string:
-        service_enable => true,
-        service_ensure => 'running',
-        port           => $port,
-        bind           => $facts['networking']['ip'],
-        dbfilename     => "${port}-dump.rdb",
-        appendfilename => "${port}-appendonly.aof",
-        appendfsync    => 'always',
-        require        => Class['Redis'],
+      class { '::redis':
+        default_install => false,
+        service_enable  => false,
+        service_ensure  => 'stopped',
       }
-    }
-
-    EOS
 
-    # Apply twice to ensure no errors the second time.
-    apply_manifest(pp, catch_failures: true)
-    apply_manifest(pp, catch_changes: true)
-  end
-
-  describe package(redis_name) do
-    it { is_expected.to be_installed }
+      $listening_ports.each |$port| {
+        $port_string = sprintf('%d',$port)
+        redis::instance { $port_string:
+          service_enable => true,
+          service_ensure => 'running',
+          port           => $port,
+          bind           => $facts['networking']['ip'],
+          dbfilename     => "${port}-dump.rdb",
+          appendfilename => "${port}-appendonly.aof",
+          appendfsync    => 'always',
+          require        => Class['Redis'],
+        }
+      }
+      PUPPET
+    end
   end
 
-  describe service(redis_name) do
-    it { is_expected.not_to be_enabled }
-    it { is_expected.not_to be_running }
-  end
+  specify { expect(package(redis_name)).not_to be_installed }
+  specify { expect(service(redis_name)).not_to be_enabled.and be_running }
```

Comment:
This line confuses me. As mentioned in the comment above - isn't `not_to ...and` disallowed in RSpec in some way?

Reply:
Alright, here it is:
```
        `expect(...).not_to matcher.and matcher` is not supported, since it creates a bit of an ambiguity. Instead, define negated versions of whatever matchers you wish to negate with `RSpec::Matchers.define_negated_matcher` and use `expect(...).to matcher.and matcher`.
```

I'd suggest either matching against a negated predicate if it exists (`disabled?` and `stopped?`) with a regular `.to` runner:
```ruby
specify { expect(service(redis_name)).to be_disabled.and be_stopped }
```
If those predicates are not available on what `service` returns, define negated matchers for `be_running` and `be_enabled`.

Or just plain split this expectation in two:
```ruby
specify do
  expect(service(redis_name)).not_to be_enabled
  expect(service(redis_name)).not_to be_running
end
```

Whatever you like the most.

Reply:
Two different tests is much easier to read and maintain.

Reply:
I suppose you meant two spec examples.
Sure, I agree, unless expensive setup is involved, which I've heard, is sometimes the case with serverspec.

Reply:
You're absolutely right that this line is vague and confusing.

It is very common to always want it both enabled and running or both disabled and stopped. I've been wondering if it makes sense to have some short hand for it, but I struggled to find a good name for it.

Looking at https://serverspec.org/resource_types.html#service there is no `be_disabled` nor `be_stopped`. I'll think a bit more about this.

Reply:
For now I went with the recommendation to split it in two statements. That was the easiest.

For example, `not_to be_running` is clear. But is `to be_stopped` the same? `to be_stopped` could imply it should be present but stopped while `not_to be_running` doesn't care whether the service is present or not. Splitting it in two avoids these issues.